lenskit-core/src/main/java/org/grouplens/lenskit/data/event/Rating.java

```diff
@@ -20,23 +20,95 @@
  */
 package org.grouplens.lenskit.data.event;
 
-import javax.annotation.Nullable;
-
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
 import org.grouplens.lenskit.data.pref.Preference;
 
+import javax.annotation.Nullable;
+import java.io.Serializable;
+
 /**
- * A rating is an expression of preference for an item by a user.
- *
- * <p>Ratings are equal if they have the same user, item, preference and timestamp.  The hash code
- * of a rating is the hash code of its preference (if it has one), or the user and item ID.
- * Timestamp is <em>ignored</em> in the hash code.  See {@link Ratings#equals(Rating, Rating)} and
- * {@link Ratings#hashRating(Rating)}.</p>
+ * A rating is an expression of preference for an item by a user.  The preference can be {@code null}, in which case
```

Comment:
I have a separate PR in the works that gets rid of the `Preference` indirection (in preparation for entirely getting rid of the `Preference` class). I thought it'd be easier to focus on one set of decisions at a time, so we can get rid of the interface business for `Rating` and then decide on ditching `Preference`.